File: scripts/shot_analysis.py
#!/usr/bin/env python3
"""Pinned, evidence-only scene-boundary analysis for final video masters.

This script intentionally has no dependency bootstrap path. The Trigger image
build installs its exact requirements into /opt/youtube-studio-qa-scene-analysis
and verifies the distributions before a task can start. A missing or mismatched
runtime is an error, not permission to invoke pip at task time.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import re
from importlib.metadata import version
from pathlib import Path
from typing import Any

SCHEMA_VERSION = "1.0.0"
SCENEDETECT_HEADLESS_VERSION = "0.7.1"
OPENCV_PYTHON_HEADLESS_VERSION = "4.12.0.88"
SHA256_RE = re.compile(r"^[a-f0-9]{64}$")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Produce a bound PySceneDetect adaptive-detector receipt."
    )
    parser.add_argument("--input", required=True, help="Final-master video to decode.")
    parser.add_argument(
        "--source-sha256",
        required=True,
        help="Expected lowercase SHA-256 of --input; mismatches fail closed.",
    )
    parser.add_argument("--adaptive-threshold", type=float, default=3.0)
    parser.add_argument("--min-scene-len-frames", type=int, default=15)
    parser.add_argument("--window-width", type=int, default=2)
    parser.add_argument("--min-content-val", type=float, default=15.0)
    return parser.parse_args()


def require_pinned_runtime() -> None:
    actual_scene_detect = version("scenedetect-headless")
    actual_opencv = version("opencv-python-headless")
    if actual_scene_detect != SCENEDETECT_HEADLESS_VERSION:
        raise RuntimeError(
            "pinned build-image dependency mismatch for scenedetect-headless: "
            f"expected {SCENEDETECT_HEADLESS_VERSION}, found {actual_scene_detect}. "
            "Runtime dependency installation is forbidden."
        )
    if actual_opencv != OPENCV_PYTHON_HEADLESS_VERSION:
        raise RuntimeError(
            "pinned build-image dependency mismatch for opencv-python-headless: "
            f"expected {OPENCV_PYTHON_HEADLESS_VERSION}, found {actual_opencv}. "
            "Runtime dependency installation is forbidden."
        )


def hash_file(path: Path) -> tuple[str, int]:
    digest = hashlib.sha256()
    byte_length = 0
    with path.open("rb") as source:
        while chunk := source.read(1024 * 1024):
            digest.update(chunk)
            byte_length += len(chunk)
    return digest.hexdigest(), byte_length


def validate_args(args: argparse.Namespace) -> tuple[Path, str]:
    source = Path(args.input)
    if not source.is_file():
        raise ValueError(f"final-master input is not a file: {source}")
    expected_hash = args.source_sha256
    if not SHA256_RE.fullmatch(expected_hash):
        raise ValueError("--source-sha256 must be exactly 64 lowercase hexadecimal characters")
    if args.adaptive_threshold <= 0:
        raise ValueError("--adaptive-threshold must be positive")
    if args.min_scene_len_frames < 1:
        raise ValueError("--min-scene-len-frames must be at least one")
    if args.window_width < 1:
        raise ValueError("--window-width must be at least one")
    if args.min_content_val <= 0:
        raise ValueError("--min-content-val must be positive")
    return source, expected_hash


def scene_receipt(args: argparse.Namespace) -> dict[str, Any]:
    source, expected_hash = validate_args(args)
    require_pinned_runtime()

    actual_hash, byte_length = hash_file(source)
    if actual_hash != expected_hash:
        raise RuntimeError(
            "final-master source hash mismatch: supplied sourceSha256 does not bind "
            "the analyzed bytes."
        )

    # Import only after exact distribution checks. `scenedetect-headless`
    # exposes the regular PySceneDetect module name, while `cv2` is import-
    # checked above to prove the headless OpenCV decoder is present.
    from scenedetect import AdaptiveDetector, detect

    detector = AdaptiveDetector(
        adaptive_threshold=args.adaptive_threshold,
        min_scene_len=args.min_scene_len_frames,
        window_width=args.window_width,
        min_content_val=args.min_content_val,
    )
    scenes = detect(str(source), detector, show_progress=False)
    serialized_scenes = [
        {
            "startFrame": int(start.get_frames()),
            "endFrameExclusive": int(end.get_frames()),
            "startSec": start.get_seconds(),
            "endSecExclusive": end.get_seconds(),
        }
        for start, end in scenes
    ]

    return {
        "schemaVersion": SCHEMA_VERSION,
        "provider": "pyscenedetect",
        "detector": "adaptive",
        "versions": {
            "scenedetectHeadless": SCENEDETECT_HEADLESS_VERSION,
            "opencvPythonHeadless": OPENCV_PYTHON_HEADLESS_VERSION,
        },
        "config": {
            "adaptiveThreshold": args.adaptive_threshold,
            "minSceneLenFrames": args.min_scene_len_frames,
            "windowWidth": args.window_width,
            "minContentVal": args.min_content_val,
        },
        "source": {"sha256": actual_hash, "byteLength": byte_length},
        "scenes": serialized_scenes,
    }


def main() -> None:
    print(json.dumps(scene_receipt(parse_args()), sort_keys=True, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: src/lib/shotAnalysis.ts
/**
 * Pinned PySceneDetect evidence adapter.
 *
 * This is deliberately a standalone analysis module. It records a reproducible
 * scene-boundary receipt for a final master, but no render or publishing gate
 * consumes it yet. Wiring a future policy consumer is a separate decision.
 */
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import { createReadStream } from "node:fs";
import { z } from "zod";

export const SHOT_ANALYSIS_SCHEMA_VERSION = "1.0.0";
export const PYSCENEDETECT_HEADLESS_VERSION = "0.7.1";
export const OPENCV_PYTHON_HEADLESS_VERSION = "4.12.0.88";
export const SHOT_ANALYSIS_SCRIPT = "scripts/shot_analysis.py";
// Do not rely on PATH: the worker also carries an isolated narration-proof
// venv. Calling the exact interpreter keeps both QA proofs available.
export const SHOT_ANALYSIS_PYTHON = "/opt/youtube-studio-qa-scene-analysis/bin/python";

const Sha256Schema = z.string().regex(/^[a-f0-9]{64}$/, "must be a lowercase SHA-256 hex digest");

export const ShotAnalysisConfigSchema = z.object({
  adaptiveThreshold: z.number().finite().positive(),
  minSceneLenFrames: z.number().int().positive(),
  windowWidth: z.number().int().positive(),
  minContentVal: z.number().finite().positive(),
}).strict();
export type ShotAnalysisConfig = z.infer<typeof ShotAnalysisConfigSchema>;

/** Adaptive detection reduces false cuts when motion or lighting varies within a shot. */
export const DEFAULT_SHOT_ANALYSIS_CONFIG: ShotAnalysisConfig = {
  adaptiveThreshold: 3.0,
  minSceneLenFrames: 15,
  windowWidth: 2,
  minContentVal: 15.0,
};

export const ShotAnalysisSceneSchema = z.object({
  startFrame: z.number().int().nonnegative(),
  endFrameExclusive: z.number().int().positive(),
  startSec: z.number().finite().nonnegative(),
  endSecExclusive: z.number().finite().positive(),
}).strict().superRefine((scene, context) => {
  if (scene.endFrameExclusive <= scene.startFrame) {
    context.addIssue({ code: z.ZodIssueCode.custom, message: "scene end frame must be after its start frame" });
  }
  if (scene.endSecExclusive <= scene.startSec) {
    context.addIssue({ code: z.ZodIssueCode.custom, message: "scene end time must be after its start time" });
  }
});
export type ShotAnalysisScene = z.infer<typeof ShotAnalysisSceneSchema>;

export const ShotAnalysisReceiptSchema = z.object({
  schemaVersion: z.literal(SHOT_ANALYSIS_SCHEMA_VERSION),
  provider: z.literal("pyscenedetect"),
  detector: z.literal("adaptive"),
  versions: z.object({
    scenedetectHeadless: z.literal(PYSCENEDETECT_HEADLESS_VERSION),
    opencvPythonHeadless: z.literal(OPENCV_PYTHON_HEADLESS_VERSION),
  }).strict(),
  config: ShotAnalysisConfigSchema,
  source: z.object({
    sha256: Sha256Schema,
    byteLength: z.number().int().positive(),
  }).strict(),
  scenes: z.array(ShotAnalysisSceneSchema),
}).strict().superRefine((receipt, context) => {
  for (let index = 1; index < receipt.scenes.length; index += 1) {
    const previous = receipt.scenes[index - 1]!;
    const current = receipt.scenes[index]!;
    if (current.startFrame < previous.endFrameExclusive || current.startSec < previous.endSecExclusive) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["scenes", index],
        message: "scene list must be chronological and non-overlapping",
      });
    }
  }
});
export type ShotAnalysisReceipt = z.infer<typeof ShotAnalysisReceiptSchema>;

export interface ShotAnalysisInvocation {
  command: typeof SHOT_ANALYSIS_PYTHON;
  args: string[];
  config: ShotAnalysisConfig;
}

export interface ShotAnalysisOptions {
  videoPath: string;
  /** Canonical SHA-256 of the exact final-master bytes expected by the caller. */
  sourceSha256: string;
  config?: Partial<ShotAnalysisConfig>;
  runner?: ShotAnalysisRunner;
  timeoutMs?: number;
}

export interface ShotAnalysisProcessResult {
  status: number | null;
  stdout: string;
  stderr: string;
  error?: Error;
}

/**
 * Hash the exact local final-master bytes before asking the baked detector to
 * decode them. This keeps the detector receipt tied to the same master final
 * QA is inspecting, without buffering a long video in memory.
 */
export async function sha256ShotAnalysisSource(videoPath: string): Promise<string> {
  if (!videoPath.trim()) throw unavailable("videoPath is required");
  const hash = createHash("sha256");
  let byteLength = 0;
  try {
    for await (const chunk of createReadStream(videoPath)) {
      const bytes = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
      hash.update(bytes);
      byteLength += bytes.byteLength;
    }
  } catch (error) {
    throw unavailable(`cannot read final-master bytes (${error instanceof Error ? error.message : String(error)})`);
  }
  if (byteLength <= 0) throw unavailable("final-master byte length is zero");
  return hash.digest("hex");
}

/** Narrow test seam; production always executes the baked Python script directly. */
export type ShotAnalysisRunner = (
  command: string,
  args: readonly string[],
  timeoutMs: number,
) => ShotAnalysisProcessResult;

function unavailable(detail: string): Error {
  return new Error(
    `shot analysis unavailable: ${detail}. The pinned build-image runtime did not complete; runtime pip fallback is forbidden.`,
  );
}

function normaliseConfig(config: Partial<ShotAnalysisConfig> | undefined): ShotAnalysisConfig {
  const parsed = ShotAnalysisConfigSchema.safeParse({ ...DEFAULT_SHOT_ANALYSIS_CONFIG, ...config });
  if (!parsed.success) {
    throw unavailable(`invalid adaptive-detector configuration (${parsed.error.issues.map((issue) => issue.message).join("; ")})`);
  }
  return parsed.data;
}

function sourceHash(value: string): string {
  const parsed = Sha256Schema.safeParse(value);
  if (!parsed.success) throw unavailable("sourceSha256 must be a lowercase SHA-256 digest");
  return parsed.data;
}

export function buildShotAnalysisInvocation(options: Pick<ShotAnalysisOptions, "videoPath" | "sourceSha256" | "config">): ShotAnalysisInvocation {
  if (!options.videoPath.trim()) throw unavailable("videoPath is required");
  const config = normaliseConfig(options.config);
  const expectedHash = sourceHash(options.sourceSha256);
  return {
    command: SHOT_ANALYSIS_PYTHON,
    args: [
      SHOT_ANALYSIS_SCRIPT,
      "--input", options.videoPath,
      "--source-sha256", expectedHash,
      "--adaptive-threshold", String(config.adaptiveThreshold),
      "--min-scene-len-frames", String(config.minSceneLenFrames),
      "--window-width", String(config.windowWidth),
      "--min-content-val", String(config.minContentVal),
    ],
    config,
  };
}

function runBakedSceneAnalyzer(command: string, args: readonly string[], timeoutMs: number): ShotAnalysisProcessResult {
  const result = spawnSync(command, args, {
    encoding: "utf8",
    maxBuffer: 8 * 1024 * 1024,
    timeout: timeoutMs,
    windowsHide: true,
  });
  return {
    status: result.status,
    stdout: result.stdout ?? "",
    stderr: result.stderr ?? "",
    ...(result.error ? { error: result.error } : {}),
  };
}

/**
 * Runs only the Python executable baked into the Trigger base image. Any
 * missing command, wrong version, invalid receipt, or source mismatch fails
 * closed; this module never installs dependencies at task time.
 */
export function analyzeShotBoundaries(options: ShotAnalysisOptions): ShotAnalysisReceipt {
  const invocation = buildShotAnalysisInvocation(options);
  const timeoutMs = options.timeoutMs ?? 20 * 60 * 1000;
  const runner = options.runner ?? runBakedSceneAnalyzer;
  const result = runner(invocation.command, invocation.args, timeoutMs);
  if (result.error || result.status !== 0) {
    const detail = result.stderr.trim().slice(-1200) || result.error?.message || `exit status ${String(result.status)}`;
    throw unavailable(`baked analyzer failed (${detail})`);
  }

  let decoded: unknown;
  try {
    decoded = JSON.parse(result.stdout);
  } catch {
    throw unavailable("baked analyzer did not emit a JSON receipt");
  }
  const parsed = ShotAnalysisReceiptSchema.safeParse(decoded);
  if (!parsed.success) {
    throw unavailable(`baked analyzer emitted an invalid receipt (${parsed.error.issues.map((issue) => issue.message).join("; ")})`);
  }
  const receipt = parsed.data;
  const expectedHash = sourceHash(options.sourceSha256);
  if (receipt.source.sha256 !== expectedHash) {
    throw unavailable("receipt source SHA-256 does not match the requested final master");
  }
  if (JSON.stringify(receipt.config) !== JSON.stringify(invocation.config)) {
    throw unavailable("receipt detector configuration does not match the requested analysis configuration");
  }
  return receipt;
}
